### PR TITLE
Utilize C++11 features in slice.hpp

### PR DESCRIPTION
**Pull request by tbeu**
See #214.

### CODE REVIEW

File: include/exiv2/slice.hpp [hunk 1]
```diff
@@ -153,22 +107,22 @@ namespace Exiv2
...
+            using const_iterator = typename storage_type<data_type>::const_iterator;
+            using value_type = typename storage_type<data_type>::value_type;
 
             /*!
              * Default contructor, requires begin to be smaller than end,
              * otherwise an exception is thrown. Also forwards all parameters to
              * the constructor of storage_
              */
-            ConstSliceBase(data_type& data, size_t begin, size_t end)
+            ConstSliceBase(data_type const& data, size_t begin, size_t end)
```
Comment (tbeu) [inline include/exiv2/slice.hpp]: This was a real change, only detected after ctor delegation.

**Comment by tbeu** [comment]
@D4N Can you please check why https://travis-ci.org/Exiv2/exiv2/jobs/491833584#L623-L644 fails?

File: include/exiv2/slice.hpp (hunk 1 above)
Comment (piponazo) [reply include/exiv2/slice.hpp]: [minor] I think in the rest of the code we are using west-const.

File: include/exiv2/slice.hpp [hunk 2]
```diff
@@ -426,9 +372,9 @@ namespace Exiv2
         template <typename storage_type>
         struct PtrSliceStorage
         {
-            typedef typename remove_cv<typename remove_pointer<storage_type>::type>::type value_type;
-            typedef value_type* iterator;
-            typedef const value_type* const_iterator;
+            using value_type = typename std::remove_cv<typename std::remove_pointer<storage_type>::type>::type;
+            using iterator = value_type*;
+            using const_iterator = const iterator;
```
Comment (D4N) [inline include/exiv2/slice.hpp]: This is the problem: `const_iterator` is not `const value_type*` as one would think (I certainly did), but it's a constant `iterator` i.e. a **constant** pointer and not a `const value_type*` which is a pointer to a constant.
```suggestion
            using const_iterator = const value_type*;
```

**Comment by D4N** [comment]
Unfortunately, that is not the full solution, the code is still broken even with the patch that I posted.

It looks like using declaration actually does things differently than just calling the constructors directly. To be honest I have currently no clue how to fix this. I remember when writing this abomination that it took me hours to correctly forward the `const` for containers & pointers and now everything is broken again. I'll give this another try tomorrow, probably sprinkling the code with `std::forward<>` might help.

**Comment by tbeu** [comment]
Superseded by #716.

**Pull request closed without merging**